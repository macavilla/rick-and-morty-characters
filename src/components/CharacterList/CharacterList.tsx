import { CharacterType } from "@/types/character";

import CharacterCard from "../CharacterCard/CharacterCard";
export interface CharacterListProps {
  characters: CharacterType[];
  listId: 1 | 2;
  selectedCharacter: CharacterType | null;
  otherSelectedCharacter: CharacterType | null;
  onSelect: (c: CharacterType | null) => void;
}

const CharacterList = ({
  listId,
  characters,
  selectedCharacter,
  otherSelectedCharacter,
  onSelect,
}: CharacterListProps) => {
  return (
    <div className="">
      <h3 className="text-2xl">Character #{listId}</h3>

      <ul className="grid grid-cols-1 sm:grid-cols-2 gap-4 border-4">
        {characters.map((character: CharacterType) => (
          <CharacterCard
            key={character.id}
            character={character}
            isSelected={character === selectedCharacter}
            isDisabled={character === otherSelectedCharacter}
            onSelect={() =>
              character !== selectedCharacter
                ? onSelect(character)
                : onSelect(null)
            }
          />
        ))}
      </ul>
    </div>
  );
};

export default CharacterList;
