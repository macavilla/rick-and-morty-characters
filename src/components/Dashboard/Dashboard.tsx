"use client";

import { useState } from "react";

import { CharacterType } from "@/types/character";

import CharacterList from "../CharacterList/CharacterList";
import Episodes from "../Episodes/Episodes";

interface DashboardProps {
  characters: CharacterType[];
}
const Dashboard = ({ characters }: DashboardProps) => {
  const [character1, setCharacter1] = useState<CharacterType | null>(null);
  const [character2, setCharacter2] = useState<CharacterType | null>(null);

  const handleResetSelection = () => {
    setCharacter1(null);
    setCharacter2(null);
  };

  return (
    <div>
      <h2>Rick and Morty Characters List:</h2>

      <button
        className="cursor-pointer bg-red-300"
        onClick={handleResetSelection}
      >
        Reset selection
      </button>
      <div className="flex justify-between w-full gap-1">
        <CharacterList
          listId={1}
          characters={characters}
          selectedCharacter={character1}
          otherSelectedCharacter={character2}
          onSelect={setCharacter1}
        />
        <CharacterList
          listId={2}
          characters={characters}
          selectedCharacter={character2}
          otherSelectedCharacter={character1}
          onSelect={setCharacter2}
        />
      </div>
      <h2>Selected Characters Epi:</h2>

      <Episodes
        character1Episode={character1?.episode || null}
        character2Episode={character2?.episode || null}
      />
    </div>
  );
};

export default Dashboard;
