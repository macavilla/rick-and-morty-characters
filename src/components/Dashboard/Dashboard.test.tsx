import { screen, render, fireEvent, within } from "@testing-library/react";
import { mockDashboardProps } from "./__mockData__/Dashboard.mock";
import { mockCharacters } from "@/components/CharacterList/__mockData__/CharacterList.mock";

import Dashboard from "./Dashboard";

jest.mock("../../utils/fetchEpisodes", () => ({
  __esModule: true,
  default: jest.fn(() =>
    Promise.resolve([
      {
        id: 1,
        name: "Pilot",
        episode: "S01E01",
        air_date: "December 2, 2013",
      },
    ])
  ),
}));

describe("Dashboard", () => {
  it("renders Dashboard and matches snapshot", () => {
    const { container } = render(<Dashboard {...mockDashboardProps} />);
    expect(container).toMatchSnapshot();
  });

  it("renders both CharacterList", () => {
    render(<Dashboard characters={mockCharacters} />);

    const lists = screen.getAllByRole("list");
    expect(lists.length).toBe(2);
  });

  it("disables Reset button initially", () => {
    render(<Dashboard characters={mockCharacters} />);
    const resetButton = screen.getByRole("button", {
      name: /reset selection/i,
    });
    expect(resetButton).toBeDisabled();
  });

  it("enables Reset button after selecting characters", () => {
    render(<Dashboard characters={mockCharacters} />);
    const resetButton = screen.getByRole("button", {
      name: /reset selection/i,
    });

    const lists = screen.getAllByRole("list");

    const list1 = lists[0];
    const list2 = lists[1];

    const character1 = within(list1).getByText("Rick Sanchez");
    const character2 = within(list2).getByText("Morty Smith");

    fireEvent.click(character1);
    fireEvent.click(character2);

    expect(resetButton).not.toBeDisabled();
  });
});
