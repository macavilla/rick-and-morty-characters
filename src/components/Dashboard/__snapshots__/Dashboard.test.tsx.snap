// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Dashboard renders Dashboard and matches snapshot 1`] = `
<div>
  <div>
    <h2>
      Rick and Morty Characters List:
    </h2>
    <button
      class="cursor-pointer bg-red-300"
    >
      Reset selection
    </button>
    <div
      class="flex justify-between w-full gap-1"
    >
      <div
        class=""
      >
        <h3
          class="text-2xl"
        >
          Character #
          1
        </h3>
        <ul
          class="grid grid-cols-1 sm:grid-cols-2 gap-4 border-4"
        >
          <div
            class="
      p-2 border rounded-md lg:flex lg:gap-1
      cursor-pointer hover:text-gray-500
      false
    "
          >
            <div
              class="w-full lg:w-1/2"
            >
              <img
                alt="Rick Sanchez"
                class="w-full h-auto"
                data-nimg="1"
                decoding="async"
                height="300"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75 2x"
                style="color: transparent;"
                width="300"
              />
            </div>
            <div
              class="lg:w-1/2"
            >
              <h4
                class="text-2xl"
              >
                Rick Sanchez
              </h4>
              <div
                class="flex items-center gap-1"
              >
                <div
                  class="rounded-full size-3 mb-0.5 bg-green-500"
                />
                <p
                  class="capitalize"
                >
                  Alive
                </p>
              </div>
              <p>
                Human
              </p>
            </div>
          </div>
          <div
            class="
      p-2 border rounded-md lg:flex lg:gap-1
      cursor-pointer hover:text-gray-500
      false
    "
          >
            <div
              class="w-full lg:w-1/2"
            >
              <img
                alt="Morty Smith"
                class="w-full h-auto"
                data-nimg="1"
                decoding="async"
                height="300"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75 2x"
                style="color: transparent;"
                width="300"
              />
            </div>
            <div
              class="lg:w-1/2"
            >
              <h4
                class="text-2xl"
              >
                Morty Smith
              </h4>
              <div
                class="flex items-center gap-1"
              >
                <div
                  class="rounded-full size-3 mb-0.5 bg-green-500"
                />
                <p
                  class="capitalize"
                >
                  Alive
                </p>
              </div>
              <p>
                Human
              </p>
            </div>
          </div>
          <div
            class="
      p-2 border rounded-md lg:flex lg:gap-1
      cursor-pointer hover:text-gray-500
      false
    "
          >
            <div
              class="w-full lg:w-1/2"
            >
              <img
                alt="Summer Smith"
                class="w-full h-auto"
                data-nimg="1"
                decoding="async"
                height="300"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F3.jpeg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F3.jpeg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F3.jpeg&w=640&q=75 2x"
                style="color: transparent;"
                width="300"
              />
            </div>
            <div
              class="lg:w-1/2"
            >
              <h4
                class="text-2xl"
              >
                Summer Smith
              </h4>
              <div
                class="flex items-center gap-1"
              >
                <div
                  class="rounded-full size-3 mb-0.5 bg-green-500"
                />
                <p
                  class="capitalize"
                >
                  Alive
                </p>
              </div>
              <p>
                Human
              </p>
            </div>
          </div>
        </ul>
      </div>
      <div
        class=""
      >
        <h3
          class="text-2xl"
        >
          Character #
          2
        </h3>
        <ul
          class="grid grid-cols-1 sm:grid-cols-2 gap-4 border-4"
        >
          <div
            class="
      p-2 border rounded-md lg:flex lg:gap-1
      cursor-pointer hover:text-gray-500
      false
    "
          >
            <div
              class="w-full lg:w-1/2"
            >
              <img
                alt="Rick Sanchez"
                class="w-full h-auto"
                data-nimg="1"
                decoding="async"
                height="300"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75 2x"
                style="color: transparent;"
                width="300"
              />
            </div>
            <div
              class="lg:w-1/2"
            >
              <h4
                class="text-2xl"
              >
                Rick Sanchez
              </h4>
              <div
                class="flex items-center gap-1"
              >
                <div
                  class="rounded-full size-3 mb-0.5 bg-green-500"
                />
                <p
                  class="capitalize"
                >
                  Alive
                </p>
              </div>
              <p>
                Human
              </p>
            </div>
          </div>
          <div
            class="
      p-2 border rounded-md lg:flex lg:gap-1
      cursor-pointer hover:text-gray-500
      false
    "
          >
            <div
              class="w-full lg:w-1/2"
            >
              <img
                alt="Morty Smith"
                class="w-full h-auto"
                data-nimg="1"
                decoding="async"
                height="300"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75 2x"
                style="color: transparent;"
                width="300"
              />
            </div>
            <div
              class="lg:w-1/2"
            >
              <h4
                class="text-2xl"
              >
                Morty Smith
              </h4>
              <div
                class="flex items-center gap-1"
              >
                <div
                  class="rounded-full size-3 mb-0.5 bg-green-500"
                />
                <p
                  class="capitalize"
                >
                  Alive
                </p>
              </div>
              <p>
                Human
              </p>
            </div>
          </div>
          <div
            class="
      p-2 border rounded-md lg:flex lg:gap-1
      cursor-pointer hover:text-gray-500
      false
    "
          >
            <div
              class="w-full lg:w-1/2"
            >
              <img
                alt="Summer Smith"
                class="w-full h-auto"
                data-nimg="1"
                decoding="async"
                height="300"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F3.jpeg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F3.jpeg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F3.jpeg&w=640&q=75 2x"
                style="color: transparent;"
                width="300"
              />
            </div>
            <div
              class="lg:w-1/2"
            >
              <h4
                class="text-2xl"
              >
                Summer Smith
              </h4>
              <div
                class="flex items-center gap-1"
              >
                <div
                  class="rounded-full size-3 mb-0.5 bg-green-500"
                />
                <p
                  class="capitalize"
                >
                  Alive
                </p>
              </div>
              <p>
                Human
              </p>
            </div>
          </div>
        </ul>
      </div>
    </div>
    <section
      class="fixed bottom-0 bg-background left-0 right-0 h-auto max-h-1/2 overflow-y-scroll"
    >
      <div
        class="md:flex md:justify-between md:gap-4 p-2 border border-t"
      >
        <div
          class="w-full md:w-1/3"
        >
          <div>
            <h4
              class="text-2xl"
            >
              Character 1 episodes
               
            </h4>
          </div>
        </div>
        <div
          class="w-full md:w-1/3"
        >
          <div>
            <h4
              class="text-2xl"
            >
              Shared episodes
               
            </h4>
          </div>
        </div>
        <div
          class="w-full md:w-1/3"
        >
          <div>
            <h4
              class="text-2xl"
            >
              Character 2 episodes
               
            </h4>
          </div>
        </div>
      </div>
    </section>
  </div>
</div>
`;
